proposal/custom-url-resolution.md

```diff
@@ -0,0 +1,247 @@
+- The new url value that we received from the url rewrite plugin than gets used to replace the original url value
+
+_Note: if there is no urlRewrite plugin/function, none of these steps should be executed._
+
+### JavaScript API
+
+At the core of remapping the sass-url references is the JavaScript API which allows users to return a new sass-url reference, based on the canonical url of the sass file and original sass-url reference.
+
+The first parameter of the urlRewrite function is an object with the following values:
+
+- `file`: The canonical url of the sass file that references sass-url.
+- `url`: The sass-url reference, for example with `sass-url(file://../assets/test.png)` it would be `file://../assets/test.png`.
+
+The second parameter of the urlRewrite function is an optional done callback that is used when performing asynchronous operations.
+
+This callback function takes in two parameters:
+
+- `error`: This can be null or an Error object, `null` means there is no error.
+- `url`: This is the `string` that gets used to replace the sass-url reference. This cannot be null or anything else, otherwise sass will throw an error.
+
+Asynchronous example:
+
+```TypeScript
+let sassOptions = {
+  // Rewrite urls asynchronously
+  rewriteUrl: async (urlReference: { file: string, url: string }, done: (error: Error | null, url: string | null) => void): void => {
+    if (urlReference.file) {
+      content = await fs.readFile(path.join(path.dirname(file), url));
+    }
+
+    done(null, content ? `data:${base64(content)))}` : url);
+  }
+}
+```
+
+Synchronous example:
+
+```TypeScript
+const outDir = '/';
+
+let sassOptions = {
+  // Rewrite urls synchronously
+  rewriteUrl: (urlReference: { file: string, url: string }): string | null => {
+    if (!urlReference.file) {
+      return url;
+    }
+
+    return path.relative(outDir, path.join(path.dirname(file), url));
+  }
+}
+```
+
+### Using Variables
+
+Variables can be used in a sass-url reference, in this case it should get remapped based on the value that gets created after the variables have been applied to ensure we are able to remap the entire url correctly.
+
+Examples with variables:
+
+```Scss
+sass-url("#{$asset-path}/image.png");
+```
+
+```Scss
+sass-url("./folder/#{$some-var}");
+```
+
+## Presets
+
+Besides having a way to allow any sass-url rewriting logic, this specification also defines 2 presets that handle common url rewriting logic: `inline` and `relative`.
+
+### The inline Preset
+
+The `inline` rewrite preset is the most fool-proof method, it inlines all references assets into the output css file using data urls.
+
+This preset should be exposed in `sass.url.inline`.
+
+#### Inline Preset Usage
+
+To use this preset you can pass an instance of this preset in the urlRewrite option.
+
+```JS
+import * as sass from 'sass';
+
+const options = {
+  rewriteUrl: sass.url.inline()
```

Comment:
should be `rewriteUrl: sass.url.inline`. We want to pass the function as the value of the option, not calling it.
There is no need to make it a function returning another function IMO.

Reply:
Right, will update